src/Test/QuickCheck/StateMachineModel/Example.hs

```diff
@@ -95,133 +124,225 @@ semStep prb (Inc ref)     = liftIO $ do
+  -- , (5, return . Untyped $ Copy STuple0)
   ]
 
-------------------------------------------------------------------------
-
-shrink1 :: Untyped MemStep ref -> [Untyped MemStep ref ]
-shrink1 (Untyped (Write ref i)) = [ Untyped (Write ref i') | i' <- shrink i ]
-shrink1 _                       = []
+returns :: MemStep resp refs -> SResponse () resp
+returns New         = SReference STuple0
+returns (Read  _)   = SResponse
+returns (Write _ _) = SResponse
+returns (Inc   _)   = SResponse
+returns (Copy _)    = SReference STuple0
+
+ixfor :: Applicative f => Proxy q -> MemStep resp p
+  -> (forall x. Sing x -> p @@ x -> f (q @@ x))
+  -> f (MemStep resp q)
+ixfor _ New             _ = pure New
+ixfor _ (Read  ref)     f = Read  <$> f STuple0 ref
+ixfor _ (Write ref val) f = Write <$> f STuple0 ref <*> pure val
+ixfor _ (Inc   ref)     f = Inc   <$> f STuple0 ref
+ixfor _ (Copy  ref)     f = Copy  <$> f STuple0 ref
 
 ------------------------------------------------------------------------
 
-instance Show ref => Show (Untyped MemStep ref) where
-  show (Untyped c) = show c
+shrink1 :: Untyped' MemStep refs -> [Untyped' MemStep refs ]
+shrink1 (Untyped' (Write ref i) iref) = [ Untyped' (Write ref i') iref | i' <- shrink i ]
+shrink1 _                               = []
 
--- https://ghc.haskell.org/trac/ghc/ticket/8128
-instance (Read ref, Typeable ref, Show ref) => Read (Untyped MemStep ref) where
+------------------------------------------------------------------------
 
+instance Read (Untyped MemStep (ConstSym1 IntRef)) where
   readPrec = parens $ do
     Ident ident <- parens lexP
     case ident of
       "New"   -> return $ Untyped New
       "Read"  -> Untyped . Read  <$> readPrec
       "Write" -> Untyped <$> (Write <$> readPrec <*> readPrec)
-      "Inc"   -> Untyped . Inc   <$> readPrec
-      _       -> empty
+      "Inc"   -> Untyped . Inc <$> readPrec
+      _        -> empty
 
   readListPrec = readListPrecDefault
 
-instance (Eq ref, Typeable ref) => Eq (Untyped MemStep ref) where
-  Untyped c1 == Untyped c2 = Just c1 == cast c2
-
-instance (Ord ref, Typeable ref) => Ord (Untyped MemStep ref) where
-  Untyped c1 <= Untyped c2 = Just c1 <= cast c2
-
-instance Functor (MemStep resp) where
-  fmap _ New           = unsafeCoerce New -- XXX: Hmm?
-  fmap f (Read  ref)   = Read  (f ref)
-  fmap f (Write ref i) = Write (f ref) i
-  fmap f (Inc   ref)   = Inc   (f ref)
-
-deriving instance Functor  (Untyped MemStep)
-deriving instance Foldable (Untyped MemStep)
+instance Read (Untyped' MemStep (ConstSym1 IntRef)) where
+  readPrec = parens $ choice
+    [ Untyped' <$ key "Untyped'" <*> parens (New <$ key " New") <*> readPrec
+    , Untyped' <$ key "Untyped'" <*> parens (Read <$ key "Read" <*> readPrec) <*> readPrec
+    , Untyped' <$ key "Untyped'" <*> parens (Write <$ key "Write" <*> readPrec <*> readPrec) <*> readPrec
+    , Untyped' <$ key "Untyped'" <*> parens (Inc <$ key "Inc" <*> readPrec) <*> readPrec
```

Comment:
We should probably remove `Copy` and `Write`, as they make the minimal cases more complicated...